spec/lib/webauthn/registration_service_spec.rb

```diff
@@ -142,10 +147,83 @@
     end
 
     it "raises a MalformedAttestationError" do
-      expect { service.register_second_factor_security_key }.to raise_error(
+      expect { service.register_security_key }.to raise_error(
         DiscourseWebauthn::MalformedAttestationError,
         I18n.t("webauthn.validation.malformed_attestation_error"),
       )
     end
   end
+
+  context "when the user presence flag is false" do
+    it "raises a UserPresenceError" do
+      # simulate missing user presence by flipping first bit to 0
+      flags = "00000010"
+      overridenAuthData = service.send(:auth_data)
+      overridenAuthData[32] = [flags].pack("b*")
+
+      service.instance_variable_set(:@auth_data, overridenAuthData)
```

Comment:
I'm curious why we need to stub here which relies on knowledge about the internal implementation instead of passing in a different set of params.

Reply:
The `user_verification` flag is set by the browser when generating the key. To pass a different set of params, we'd need to generate, store and pass a whole new key. And doing so would be more opaque to future travellers, the contents of a key is by design not easy to inspect. 

In this case, fiddling with the internal implementation is convenient and clearer to readers. We know the key being used is valid (from the earlier test case) and we can easily see that a valid key with a missing verification flag fails.